### ISSUE TITLE
Static analyzer: suppress warning for 3rd-party/system libs

### ISSUE THREAD

**Issue by abeimler** [aminya/project_options#133]
Is it possible to exclude targets or 3rd-party libs from the static analyzer ?

**Root Project**
```cmake
option(ENABLE_TESTING "Enable the tests" ${PROJECT_IS_TOP_LEVEL})
if(ENABLE_TESTING)
  set(ENABLE_CLANG_TIDY "ENABLE_CLANG_TIDY")
  set(ENABLE_CPPCHECK "ENABLE_CPPCHECK")
  set(ENABLE_COVERAGE "ENABLE_COVERAGE")

  if(NOT "${CMAKE_SYSTEM_NAME}" STREQUAL "Windows")
    set(ENABLE_SANITIZER_ADDRESS "ENABLE_SANITIZER_ADDRESS")
    set(ENABLE_SANITIZER_UNDEFINED_BEHAVIOR "ENABLE_SANITIZER_UNDEFINED_BEHAVIOR")
  else()
    # or it is MSVC and has run vcvarsall
    string(FIND "$ENV{PATH}" "$ENV{VSINSTALLDIR}" index_of_vs_install_dir)
    if(MSVC AND "${index_of_vs_install_dir}" STREQUAL "-1")
      set(ENABLE_SANITIZER_ADDRESS "ENABLE_SANITIZER_ADDRESS")
    endif()
  endif()
endif()

# defaulted_project_options sets recommended defaults and provides user and developer
# modes and full GUI support for choosing options at configure time
# for more flexibility, look into project_options() macro
# Initialize project_options variable related to this project
# This overwrites `project_options` and sets `project_warnings`
# uncomment the options to enable them:
dynamic_project_options(
  ENABLE_CACHE
  ${ENABLE_CPPCHECK}
  ${ENABLE_CLANG_TIDY}
  ENABLE_VS_ANALYSIS
  ${ENABLE_COVERAGE}
  ${ENABLE_SANITIZER_ADDRESS}
  ${ENABLE_SANITIZER_UNDEFINED_BEHAVIOR}
  # Note: PCH is disabled by default in developer mode because these headers become
  # globally included and they can mask other errors
  PCH_HEADERS
  # This is a list of headers to pre-compile, here are some common ones
  <vector>
  <string>
  <utility>
  <array>
  <algorithm>
  <concepts>
  CPPCHECK_OPTIONS
  --enable=style,performance,warning,portability
  --inline-suppr
  # We cannot act on a bug/missing feature of cppcheck
  --suppress=cppcheckError
  --suppress=internalAstError
  # if a file does not have an internalAstError, we get an unmatchedSuppression error
  --suppress=unmatchedSuppression
  --suppress=passedByValue
  --suppress=syntaxError
  --inconclusive
)
target_compile_features(project_options INTERFACE cxx_std_${CMAKE_CXX_STANDARD})
```

**Add lib cmake**
```cmake
# get raylib from github and build as static lib
cpmaddpackage(
    NAME
    raylib
    GIT_TAG
    4.0.0
    GITHUB_REPOSITORY
    raysan5/raylib
    OPTIONS
    "BUILD_SHARED_LIBS OFF")
if(NOT MSVC)
  # don't error on implicit declarations, which are invalid in C99 but commonly used
  target_compile_options(raylib PRIVATE -Wno-error=implicit-function-declaration)
endif()
```

**Project cmake**
```cmake
add_library(engine STATIC)
# more library options
target_link_libraries(engine PRIVATE project_warnings project_options)

# add 3rd-party libs as system libs
target_link_system_libraries(engine PUBLIC raylib)
```

**Warnings from raylib**
```bash
[2/127] Building C object _deps/raylib-build/raylib/CMakeFiles/raylib.dir/rtextures.c.o
Checking cmake-build-debug/_deps/raylib-src/src/rtextures.c ...
Checking cmake-build-debug/_deps/raylib-src/src/rtextures.c: GRAPHICS_API_OPENGL_33=1;PLATFORM_DESKTOP=1...
cmake-build-debug/_deps/raylib-src/src/external/stb_image.h:7013:37: warning: Array 'signature[8]' accessed at index 10, which is out of bounds. [arrayIndexOutOfBounds]
      if (stbi__get8(s) != signature[i])
                                    ^
cmake-build-debug/_deps/raylib-src/src/external/stb_image_write.h:1378:16: warning: Array index 'i' is used before limits check. [arrayIndexThenCheck]
      for (; DU[i]==0 && i<=end0pos; ++i) {
               ^
cmake-build-debug/_deps/raylib-src/src/external/stb_image.h:4316:23: warning: Condition 'c==18' is always true [knownConditionTrueFalse]
         } else if (c == 18) {
                      ^
cmake-build-debug/_deps/raylib-src/src/external/stb_image.h:4314:23: note: Assuming that condition 'c==17' is not redundant
         } else if (c == 17) {
                      ^
cmake-build-debug/_deps/raylib-src/src/external/stb_image.h:4305:22: note: Assuming that condition 'c>=19' is not redundant
      if (c < 0 || c >= 19) return stbi__err("bad codelengths", "Corrupt PNG");
```

**environment**

* **OS:** Linux
* **Compiler:** gcc 11
* **Cmake:** cmake 3.22.1
* **IDE:** CLion

**Comment by ClausKlein** [comment]
I would always use `run-clang-tidy` explizit!

If you enable 
```cmake
  ${ENABLE_CPPCHECK}
  ${ENABLE_CLANG_TIDY}
```
the build takes too long time and you have no control which files will be checked.

## To prevent warnings from 3rd party libs you have to options:
1.) install them first and use `find_package()`
2.) or do it like [here](https://github.com/ClausKlein/ModernCmakeStarter/blob/531bf08293a624e35f96d187677e9f8a5460847e/GNUmakefile#L46)

**Comment by abeimler** [comment]
> I would always use `run-clang-tidy` explizit!
> 
> If you enable
> 
> ```cmake
>   ${ENABLE_CPPCHECK}
>   ${ENABLE_CLANG_TIDY}
> ```
> 
> the build takes too long time and you have no control which files will be checked.

hmm, yes I may use it in the CI ... (but the problem still remains)

> ## To prevent warnings from 3rd party libs you have to options:
> 1.) install them first and use `find_package()` 2.) or do it like [here](https://github.com/ClausKlein/ModernCmakeStarter/blob/531bf08293a624e35f96d187677e9f8a5460847e/GNUmakefile#L46)

... but I just want to use what I have.
While I agree with the `find_package()`, this just works for package managed or other system dependencies, but not for dependencies building from scratch.

**Comment by abeimler** [comment]
I try to set the clang-tidy property for the target, but it didn't work:
```cmake
set_target_properties(raylib PROPERTIES CXX_CLANG_TIDY "")
```

For `cppcheck`, just use the `CPPCHECK_OPTIONS`:
```cmake
  --suppress=*:${CMAKE_BINARY_DIR}/_deps/*
  --suppress=*:${CMAKE_SOURCE_DIR}/libs/*
```

**Comment by fgallegosalido** [comment]
> I try to set the clang-tidy property for the target, but it didn't work:
> 
> ```cmake
> set_target_properties(raylib PROPERTIES CXX_CLANG_TIDY "")
> ```
> 
> For `cppcheck`, just use the `CPPCHECK_OPTIONS`:
> 
> ```cmake
>   --suppress=*:${CMAKE_BINARY_DIR}/_deps/*
>   --suppress=*:${CMAKE_SOURCE_DIR}/libs/*
> ```

Setting the clang-tidy property for the target to an empty string works for me. Are you sure the output you see doesn't come from the compiler? I'm using the following macro:

```cmake
macro(TargetDisableClangTidy target)
    set_target_properties(${target}
        PROPERTIES
            CXX_CLANG_TIDY ""
    )
endmacro()
```

**Comment by abeimler** [comment]
After some testing ...  
I figure it out, I used the wrong property lol (it's a **C** library)

```cmake
macro(DisableStaticAnalysisForTarget target)
  set_target_properties(${target} PROPERTIES C_CPPCHECK "")
  set_target_properties(${target} PROPERTIES C_CLANG_TIDY "")
  set_target_properties(${target} PROPERTIES CXX_CPPCHECK "")
  set_target_properties(${target} PROPERTIES CXX_CLANG_TIDY "")
endmacro()



add_subdirectory(mylibexample)
DisableStaticAnalysisForTarget(MyCppLibExample)
DisableStaticAnalysisForTarget(MyCLibExample)
```

**Comment by aminya** [comment]
@abeimler We can add this function to the project if you want

**Comment by aminya** [comment]
Fixed in #134
